File: src/javascript/algorithms/tree/levelOrderTraversal.js
//BFS: TC: O(n) SC: O(n)

function levelOrder(root) {
    let traversal = [];
    let queue = [];
    queue.push(root);

    while(queue.length) {
        let len = queue.length;
        let currLevel = [];

        while(len >0){
            let node = queue.shift();
            currLevel.push(node.value);
            if(node.left !== null) queue.push(node.left);
            if(node.right !== null) queue.push(node.right);
            len--;
        }
        traversal.push(currLevel);
    }
    return traversal;
}

class TreeNode {
    constructor(value) {
        this.left = null;
        this.right = null;
        this.value = value;
    }
}

let root = new TreeNode(-10);
root.left = new TreeNode(9);
root.right = new TreeNode(20);
root.right.left = new TreeNode(15);
root.right.right = new TreeNode(7);

console.log(levelOrder(root));

File: src/java1/algorithms/tree/levelOrderTraversal/LevelOrderTraversal.java
package java1.algorithms.tree.levelOrderTraversal;
import java.util.*;

import java1.algorithms.tree.TreeNode;

public class LevelOrderTraversal {
    //BFS: TC: O(n) SC: O(n)
    private static List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> traversal = new ArrayList<>();
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);

        while (!queue.isEmpty()) {
            int size = queue.size();
            List<Integer> currLevel = new ArrayList<>();
            while(size > 0) {
                TreeNode node = queue.poll();
                if(node != null) {
                    currLevel.add(node.value);
                    if(node.left != null) queue.add(node.left);
                    if(node.right != null) queue.add(node.right);
                }
                size--;
            }
            if(currLevel.size() > 0) {
                traversal.add(currLevel);
            }
        }

        return traversal;
    }

    public static void main(String[] args) {
        TreeNode root1 = new TreeNode(1);
        root1.left = new TreeNode(2);
        root1.right = new TreeNode(3);
        root1.left.left = new TreeNode(4);
        root1.left.right = new TreeNode(5);
        root1.right.left = new TreeNode(6);
        root1.right.right = new TreeNode(7);

        TreeNode root2 = new TreeNode(3);

        System.out.println(levelOrder(root1));
        System.out.println(levelOrder(root2));
        System.out.println(levelOrder(null));
    }
}
